modules/garbage.py

```diff
@@ -2,124 +2,108 @@
             yield self.env.timeout(self.sleep)  # wait for...
 
-
     def collect(self):
-        print("At %.3f, CGC GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
-
+        print("At %.3f, GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
+        self.collect_exe = True
+        self.server.action.interrupt()
         while self.heap.level > 0:                                          # while threshold is empty...
             trash = self.heap.level                                         # keeps the amount of trash
             yield self.env.timeout(self.gc_execution_time_by_trash(trash))  # run the time of discarting
             yield self.heap.get(trash)                                      # discards the trash
 
-        print("At %.3f, CGC GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
+        self.collect_exe = False
+        print("At %.3f, GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
+        self.server.action = self.env.process(self.server.run())
 
     def gc_execution_time_by_trash(self, trash):
         """ implement the way to calculate the execution time of Garbage Collector """
         return trash
 
-class STWGC(GC):
-
-    def __init__(self, env, server, threshold=1, sleep=2):
-        super().__init__(env, server.heap, threshold, sleep)
-        self.server = server
-
-    def collect(self):
-        print("At %.3f, STWGC GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
-
-        self.server.action.interrupt()
-        if self.server.gc.collect_exe:
-            self.server.gc.collect_exe.interrupt()
-
-        while self.heap.level > 0:                                          # while threshold is empty...
-            trash = self.heap.level                                         # keeps the amount of trash
-            yield self.env.timeout(self.gc_execution_time_by_trash(trash))  # run the time of discarting
-            yield self.heap.get(trash)                                      # discards the trash
-
-        self.server.action = self.env.process(self.server.run())
-
-        print("At %.3f, STWGC GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
 
 class GCI(object):
 
-    def __init__(self, env,  threshold=0.7, check_heap=2, initial_gc_exec_time=0.200):
+    def __init__(self, env, server,  threshold=0.7, check_heap=2, initial_gc_exec_time=0, sleep=0.00002):
         self.env = env
+        self.server = server
         self.threshold = threshold
         self.check_heap = check_heap
         self.gc_exec_time = initial_gc_exec_time
+        self.sleep = sleep
 
         self.shed_requests = False
-
         self.processed_requests_history = list()
         self.gc_execution_history = list()
         self.history_size = 5
+        self.times_performed = 0
 
-        self.sleep = 0.02
+    def check(self):
```

Comment:
Maybe I am missing this flow, but I believe there is a check missing:

* What if a request arrives and the server is unavailable (for instance, the GC is already running)? This request should be shed.

Reply:
Yes. This logical is implemented at server. The method request_arrived call that check() for each request received. After that, the server check if the GCI is in shed. If it is in shed, he refuse. Otherwise, the server put it in the queue.

Reply:
Would be super nice if we could consolidate this flow. Any ideas, @dfquaresma ? I totally understand if it is not possible.

Reply:
I have created a method called intercept in GCI that received the request and do what os been done at server. Feel free do disagree, I may think on another solution.